core/services/bulletprooftxmanager/eth_confirmer.go

```diff
@@ -269,32 +302,34 @@ func (ec *ethConfirmer) handleAnyInProgressAttempts(ctx context.Context, blockHe
 	return nil
 }
 
-func getInProgressEthTxAttempts(s *store.Store) ([]models.EthTxAttempt, error) {
+func getInProgressEthTxAttempts(s *store.Store, address gethCommon.Address) ([]models.EthTxAttempt, error) {
 	var attempts []models.EthTxAttempt
 	err := s.DB.
 		Preload("EthTx").
 		Joins("INNER JOIN eth_txes ON eth_txes.id = eth_tx_attempts.eth_tx_id AND eth_txes.state in ('confirmed', 'unconfirmed')").
 		Where("eth_tx_attempts.state = 'in_progress'").
+		Where("eth_txes.from_address = ?", address).
 		Find(&attempts).Error
 	return attempts, errors.Wrap(err, "getInProgressEthTxAttempts failed")
 }
 
 // FindEthTxsRequiringNewAttempt returns transactions that have all
-// attempts which are unconfirmed for at least gasBumpThreshold blocks
-func FindEthTxsRequiringNewAttempt(db *gorm.DB, blockNum int64, gasBumpThreshold int64) ([]models.EthTx, error) {
-	var etxs []models.EthTx
-	err := db.
+// attempts which are unconfirmed for at least gasBumpThreshold blocks,
+// limited by limit pending transactions
+func FindEthTxsRequiringNewAttempt(db *gorm.DB, address gethCommon.Address, blockNum, gasBumpThreshold, limit int64) (etxs []models.EthTx, err error) {
+	err = db.
 		Preload("EthTxAttempts", func(db *gorm.DB) *gorm.DB {
 			return db.Order("eth_tx_attempts.gas_price DESC")
 		}).
 		Joins("LEFT JOIN eth_tx_attempts ON eth_txes.id = eth_tx_attempts.eth_tx_id "+
 			"AND eth_tx_attempts.state != 'insufficient_eth' "+
 			"AND (broadcast_before_block_num > ? OR broadcast_before_block_num IS NULL OR eth_tx_attempts.state != 'broadcast')", blockNum-gasBumpThreshold).
-		Order("nonce ASC").
 		Where("eth_txes.state = 'unconfirmed' AND eth_tx_attempts.id IS NULL").
+		Where("eth_txes.id IN (SELECT id FROM eth_txes WHERE state = 'unconfirmed' AND from_address = ? ORDER BY nonce ASC LIMIT ?)", address, limit).
```

Comment:
If you have 20 unconfirmed transactions, a tx bump depth of 5, and bump pending transactions after 3 blocks: won't this end up actually bumping 15 transactions? First time around it bumps 5, and then on the next block, those 5 don't meet the broadcast_before_block_num constraint, so you bump the next 5, another 5 on the third block. On the fourth block you'd circle back and bump the original 5 again. This seems like extra mental gymnastics for node operators, and it would be nice to just give them a setting that says bump 5, or 15, but not have to take ETH_GAS_BUMP_THRESHOLD into account.

Reply:
Not quite.

The subquery can be imagined to execute entirely separately from the rest, and the subquery is quite simple:

`SELECT id FROM eth_txes WHERE state = 'unconfirmed' AND from_address = ? ORDER BY nonce ASC LIMIT <gas bump threshold>`

This always selects the N transations at the bottom of the pending list, regardless of their gas bumping status.

You can imagine it as a mask we place on top of the previous query, that filters out any transactions above N in the pending stack. I think this was the original intent of the PR, no?